From a2c0b852a0f4c51558dcbb7bc21c99ef91c5551a Mon Sep 17 00:00:00 2001
From: Harald Hoyer <harald@redhat.com>
Date: Thu, 18 Jul 2013 12:54:11 +0200
Subject: [PATCH] Install multipath module only, when root is multipath in
 generic mode

Even in the generic mode, we don't want to install the multipath module,
if the root device is not multipathed, because it would always have to
be synced with the real root.

Also install /etc/multipath-root.conf /etc/multipath-root/*
/etc/xdrdevices-root.conf, if present, in the initramfs as the files
without "-root", taking priority over the normal files.

https://bugzilla.redhat.com/show_bug.cgi?id=916144
---
 modules.d/90multipath/check   | 12 ++++--------
 modules.d/90multipath/install | 27 ++++++++++++++++++++++-----
 2 files changed, 26 insertions(+), 13 deletions(-)

diff --git a/modules.d/90multipath/check b/modules.d/90multipath/check
index b1834b5..984ac98 100755
--- a/modules.d/90multipath/check
+++ b/modules.d/90multipath/check
@@ -14,12 +14,8 @@ is_mpath() {
     return 1
 }
 
-if [[ $1 = -h ]]; then
-    rootdev=$(find_root_block_device)
-    if [[ $rootdev ]]; then
-        check_block_and_slaves is_mpath "$rootdev" && exit 0
-    fi
-    exit 1
+rootdev=$(find_root_block_device)
+if [[ $rootdev ]]; then
+    check_block_and_slaves is_mpath "$rootdev" && exit 0
 fi
-
-exit 0
\ No newline at end of file
+exit 1
diff --git a/modules.d/90multipath/install b/modules.d/90multipath/install
index c7774e9..56b1307 100755
--- a/modules.d/90multipath/install
+++ b/modules.d/90multipath/install
@@ -1,11 +1,32 @@
 #!/bin/bash
-
 if ldd $(which multipath) 2>/dev/null |grep -q lib64; then
     LIBDIR="/lib64"
 else
     LIBDIR="/lib"
 fi
 
+if [[ -f /etc/xdrdevices-root.conf ]]; then
+    inst_simple /etc/xdrdevices-root.conf /etc/xdrdevices.conf
+else
+    [[ -f /etc/xdrdevices.conf ]] && inst_simple /etc/xdrdevices.conf
+fi
+
+if [[ -f /etc/multipath-root.conf ]]; then
+    inst_simple /etc/multipath-root.conf /etc/multipath.conf
+else
+    [[ -f /etc/multipath.conf ]] && inst_simple /etc/multipath.conf
+fi
+
+if [[ -d /etc/multipath-root ]]; then
+    for f in /etc/multipath-root/* ;do
+        [ -e "$f" ] && inst_simple "$f" /etc/multipath/${f##*/}
+    done
+else
+    for f in /etc/multipath/* ;do
+        [ -e "$f" ] && inst_simple "$f"
+    done
+fi
+
 for f in  \
     /sbin/dmsetup \
     /sbin/kpartx \
@@ -14,9 +35,6 @@ for f in  \
     /sbin/multipathd \
     /sbin/xdrgetuid \
     /sbin/xdrgetprio \
-    /etc/xdrdevices.conf \
-    /etc/multipath.conf \
-    /etc/multipath/* \
     $(ls $LIBDIR/libmultipath* $LIBDIR/multipath/* 2>/dev/null) \
     	;do
     [ -e "$f" ] && inst "$f"
@@ -25,4 +43,3 @@ done
 inst_hook pre-trigger 02 "$moddir/multipathd.sh"
 inst_hook pre-pivot   02 "$moddir/multipathd-stop.sh"
 inst_rules 40-multipath.rules
-
